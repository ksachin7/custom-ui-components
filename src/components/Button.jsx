import React from 'react'; // eslint-disable-line no-unused-vars
import PropTypes from 'prop-types';
import { styled, css } from 'styled-components';

const variations = {
  contained: css`
    color: #fff;
    background-color: ${props => getColor(props.color)};
    border: none;
    &:hover {
      background-color: ${props => getHoverColor(props.color)};
    }
  `,
  outlined: css`
    color: ${props => getColor(props.color)};
    background-color: transparent;
    border: 1px solid ${props => getColor(props.color)};
    &:hover {
      border-color: ${props => getHoverColor(props.color)};
      color: ${props => getHoverColor(props.color)};
    }
  `,
  text: css`
    color: var(--color-grey-500);
    background-color: transparent;
    border: none;
    &:hover {
      color: ${props => getHoverColor(props.color)};
    }
  `
};

const sizes = {
  sm: css`
    font-size: 14px;
    padding: 6px 12px;
  `,
  md: css`
    font-size: 16px;
    padding: 10px 20px;
  `,
  lg: css`
    font-size: 18px;
    padding: 14px 28px;
  `,
};

const colors = {
  primary: 'var(--color-blue-600)',
  secondary: 'var(--color-grey-600)',
  info: 'var(--color-sky-500)',
  success: 'var(--color-green-500)',
  warning: 'var(--color-yellow-500)',
  danger: 'var(--color-red-700)',
};

const getColor = (color) => {
  return colors[color] || colors.primary;
};

const getHoverColor = (color) => {
  switch (color) {
    case 'primary':
      return 'var(--color-blue-700)';
    case 'secondary':
      return 'var(--color-grey-700)';
    case 'info':
      return 'var(--color-sky-700)';
    case 'success':
      return 'var(--color-green-700)';
    case 'warning':
      return 'var(--color-yellow-700)';
    case 'danger':
      return 'var(--color-red-800)';
    case 'text':
      return 'transparent';
    default:
      return 'var(--color-blue-700)';
  }
};

const StyledButton = styled.button`
  border-radius: 4px;
  transition: background-color 0.3s, color 0.3s, border-color 0.3s;
  ${props => variations[props.variant]}
  ${props => sizes[props.size]}
  /* text-transform: ${props => props.uppercase ? 'uppercase' : 'none'}; */
  cursor: ${props => props.disabled ? 'not-allowed' : 'pointer'};
  /* pointer-events: ${props => props.disabled ? 'none' : 'auto'}; */
  &:focus{
    outline: none;
  } 
  &:disabled {
    background-color: #f5f5f5;
    border: 1px solid #e0e0e0;
    color: #a0a0a0;
    box-shadow: none;
  }
  &:disabled:hover {
    background-color: #f5f5f5;
    border: 1px solid #e0e0e0;
    color: #a0a0a0;
    box-shadow: none;
  }
`;

function Button({ children, variant, color, size, disabled, uppercase, ...otherProps }) {
  return (
    <StyledButton type="button" variant={variant} color={color} size={size} {...otherProps} disabled={disabled ? true : false} style={{ textTransform: uppercase !== false ? 'uppercase' : 'none' }}>
      {children}
    </StyledButton>
  );
}

Button.propTypes = {
  variant: PropTypes.string,
  color: PropTypes.string,
  size: PropTypes.string,
  uppercase: PropTypes.bool,
  disabled: PropTypes.bool,
  children: PropTypes.node.isRequired, 
}

Button.defaultProps = {
  variant: 'contained',
  color: 'primary',
  size: 'md',
  uppercase: false,
};

export default Button;
