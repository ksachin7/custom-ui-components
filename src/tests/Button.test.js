import React from 'react'; // eslint-disable-line no-unused-vars
import { mount } from 'enzyme';
import Button from '../components/Button';

describe('Button component', () => {
  it('renders with default props', () => {
    const wrapper = mount(<Button>Default Button</Button>);
    const button = wrapper.find('button');
    expect(button.prop('type')).toBe('button');
    expect(button.prop('disabled')).toBe(false);
    // expect(wrapper).toMatchSnapshot();
  });  

  it('renders with custom props', () => {
    const wrapper = mount(
      <Button variant="outlined" color="secondary" size="lg" uppercase>
        Custom Button
      </Button>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('calls onClick callback when clicked', () => {
    const onClickMock = jest.fn();
    const wrapper = mount(<Button onClick={onClickMock}>Clickable Button</Button>);
    wrapper.find('button').simulate('click');
    expect(onClickMock).toHaveBeenCalled();
  });
});
