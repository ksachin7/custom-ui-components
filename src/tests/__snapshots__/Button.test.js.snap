// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component renders with custom props 1`] = `ReactWrapper {}`;
